File: Sources/Bifrost/Services/GameLocator.swift
import Foundation

/// Locates the Valheim install through Steam's own bookkeeping. We
/// deliberately never guess at well-known install paths (there's a stray
/// `~/Applications/Valheim.app` on this machine that must never be used) —
/// the only trustworthy source of truth is Steam's library folder list and
/// app manifest.
enum GameLocator {
    static let valheimAppID = "892970"

    struct LocatedGame: Sendable, Equatable {
        let directory: URL
        let hasExecutable: Bool

        var isValid: Bool { hasExecutable }
    }

    enum LocatorError: Error {
        case steamNotFound
        case libraryFoldersNotFound
        case appNotOwned
    }

    static var steamRoot: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/Steam")
    }

    /// Finds the Valheim install directory by walking Steam's configured
    /// library folders and looking for an app manifest for 892970.
    static func locate() -> LocatedGame? {
        let libraries = steamLibraryPaths()
        for library in libraries {
            let steamappsDir = library.appendingPathComponent("steamapps")
            let manifestURL = steamappsDir.appendingPathComponent("appmanifest_\(valheimAppID).acf")
            guard let manifestContents = try? String(contentsOf: manifestURL, encoding: .utf8) else {
                continue
            }

            let installDir = vdfValue(for: "installdir", in: manifestContents) ?? "Valheim"
            let gameDir = steamappsDir
                .appendingPathComponent("common")
                .appendingPathComponent(installDir)

            let appBundle = gameDir.appendingPathComponent("valheim.app")
            let hasExecutable = FileManager.default.fileExists(atPath: appBundle.path)
            return LocatedGame(directory: gameDir, hasExecutable: hasExecutable)
        }
        return nil
    }

    /// Reads `libraryfolders.vdf` and returns every configured library path,
    /// including the Steam root itself (which is always an implicit
    /// library). Tolerates both the older flat format
    /// (`"1" "/some/path"`) and the newer nested-object format
    /// (`"1" { "path" "/some/path" ... }`).
    static func steamLibraryPaths() -> [URL] {
        let root = steamRoot
        var paths: [URL] = [root]

        let libraryFoldersURL = root
            .appendingPathComponent("steamapps")
            .appendingPathComponent("libraryfolders.vdf")

        guard let contents = try? String(contentsOf: libraryFoldersURL, encoding: .utf8) else {
            return paths
        }

        for line in contents.components(separatedBy: .newlines) {
            guard let path = vdfLineValue(for: "path", in: line) else { continue }
            let url = URL(fileURLWithPath: path)
            if !paths.contains(url) {
                paths.append(url)
            }
        }

        return paths
    }

    /// Extracts the value for `"key"    "value"` from a single VDF line.
    private static func vdfLineValue(for key: String, in line: String) -> String? {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard trimmed.hasPrefix("\"\(key)\"") else { return nil }

        // Strip the key, then pull the next quoted token.
        let afterKey = trimmed.dropFirst(key.count + 2)
        return firstQuotedToken(in: String(afterKey))
    }

    /// Extracts the value for `"key"    "value"` anywhere in a multi-line
    /// VDF blob (used for flat manifest files like appmanifest_*.acf).
    private static func vdfValue(for key: String, in contents: String) -> String? {
        for line in contents.components(separatedBy: .newlines) {
            if let value = vdfLineValue(for: key, in: line) {
                return value
            }
        }
        return nil
    }

    private static func firstQuotedToken(in text: String) -> String? {
        guard let firstQuote = text.firstIndex(of: "\"") else { return nil }
        let afterFirst = text.index(after: firstQuote)
        guard let secondQuote = text[afterFirst...].firstIndex(of: "\"") else { return nil }
        return String(text[afterFirst..<secondQuote])
    }

    // MARK: - BepInEx

    /// A BepInEx install is considered present when the loader's directory
    /// tree and doorstop shim are all sitting alongside the game.
    static func bepinexInstalled(at gameDir: URL) -> Bool {
        let fm = FileManager.default
        let markers = [
            "BepInEx",
            "doorstop_libs",
            "doorstop_config.ini",
            "start_game_bepinex.sh",
        ]
        return markers.allSatisfy { fm.fileExists(atPath: gameDir.appendingPathComponent($0).path) }
    }

    // MARK: - Steam launch options

    /// Checks whether any Steam user profile has launch options for
    /// Valheim (892970) that route through Bifrost's `run_modded.sh`
    /// wrapper. This is a targeted scan, not a full VDF parse: it finds the
    /// `"892970" { ... }` block nested under `"apps"` and looks for a
    /// `LaunchOptions` value mentioning our wrapper script.
    static func steamConfiguredForModdedLaunch() -> Bool {
        let userdataRoot = steamRoot.appendingPathComponent("userdata")
        guard let profiles = try? FileManager.default.contentsOfDirectory(
            at: userdataRoot,
            includingPropertiesForKeys: [.contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else {
            return false
        }

        // Prefer the most recently modified profile, but fall back to
        // checking all of them if that one doesn't have it configured.
        let sortedProfiles = profiles.sorted { lhs, rhs in
            let lhsDate = (try? lhs.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? .distantPast
            let rhsDate = (try? rhs.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? .distantPast
            return lhsDate > rhsDate
        }

        for profile in sortedProfiles {
            let configURL = profile.appendingPathComponent("config/localconfig.vdf")
            guard let contents = try? String(contentsOf: configURL, encoding: .utf8) else { continue }
            if launchOptionsReferenceRunModded(in: contents) {
                return true
            }
        }
        return false
    }

    /// Scans for the `"892970" { ... "LaunchOptions" "..." ... }` block
    /// under `"apps"` and checks that its LaunchOptions value mentions
    /// run_modded.sh. Line-based and deliberately tolerant of the rest of
    /// the file's structure.
    private static func launchOptionsReferenceRunModded(in contents: String) -> Bool {
        let lines = contents.components(separatedBy: .newlines)
        guard let appIDLineIndex = lines.firstIndex(where: { line in
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            return trimmed == "\"\(valheimAppID)\""
        }) else {
            return false
        }

        // Scan forward from the app id line for a LaunchOptions entry,
        // stopping once we've closed that app's brace block.
        var depth = 0
        var sawOpenBrace = false
        for line in lines[appIDLineIndex...] {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed == "{" {
                depth += 1
                sawOpenBrace = true
                continue
            }
            if trimmed == "}" {
                depth -= 1
                if sawOpenBrace && depth <= 0 { break }
                continue
            }
            if trimmed.hasPrefix("\"LaunchOptions\"") {
                return trimmed.contains("run_modded.sh")
            }
        }
        return false
    }
}

File: src/Bifrost.Core/Services/GameLocator.cs
namespace Bifrost.Core.Services;

/// <summary>
/// Locates the Valheim install through Steam's own bookkeeping — the
/// registry-resolved (or overridden) Steam root, its
/// <c>steamapps/libraryfolders.vdf</c>, and the app manifest for 892970.
/// Ported from the macOS reference implementation's <c>GameLocator.swift</c>.
///
/// Every path this walks is overridable (constructor parameter or the
/// <c>BIFROST_STEAM_ROOT</c> environment variable — see
/// <see cref="BifrostPaths.ResolveSteamRoot"/>), so <c>--check</c> can drive
/// this entirely against fixtures on a non-Windows dev machine.
/// </summary>
public sealed class GameLocator
{
    public const string ValheimAppId = "892970";

    public sealed record LocatedGame(string Directory, bool HasExecutable)
    {
        public bool IsValid => HasExecutable;
    }

    private readonly string _steamRoot;

    public GameLocator(string? steamRoot = null)
    {
        _steamRoot = steamRoot ?? BifrostPaths.ResolveSteamRoot();
    }

    public string SteamRoot => _steamRoot;

    /// <summary>
    /// Every configured Steam library path, including the Steam root itself
    /// (always an implicit library).
    /// </summary>
    public List<string> SteamLibraryPaths()
    {
        var paths = new List<string> { _steamRoot };

        var libraryFoldersPath = Path.Combine(_steamRoot, "steamapps", "libraryfolders.vdf");
        if (!File.Exists(libraryFoldersPath))
        {
            return paths;
        }

        var contents = TryReadAllText(libraryFoldersPath);
        if (contents is null)
        {
            return paths;
        }

        foreach (var path in VdfParser.ParseLibraryFolderPaths(contents))
        {
            if (!paths.Contains(path))
            {
                paths.Add(path);
            }
        }
        return paths;
    }

    /// <summary>
    /// Finds the Valheim install directory by walking every configured
    /// library and looking for an app manifest for 892970.
    /// </summary>
    public LocatedGame? Locate()
    {
        foreach (var library in SteamLibraryPaths())
        {
            var steamappsDir = Path.Combine(library, "steamapps");
            var manifestPath = Path.Combine(steamappsDir, $"appmanifest_{ValheimAppId}.acf");
            var manifestContents = TryReadAllText(manifestPath);
            if (manifestContents is null)
            {
                continue;
            }

            var installDir = VdfParser.GetValue("installdir", manifestContents) ?? "Valheim";
            var gameDir = Path.Combine(steamappsDir, "common", installDir);

            var exePath = Path.Combine(gameDir, "valheim.exe");
            var hasExecutable = File.Exists(exePath);
            return new LocatedGame(gameDir, hasExecutable);
        }
        return null;
    }

    /// <summary>
    /// A BepInEx install is considered present when the loader's core
    /// directory, its Windows doorstop shim (winhttp.dll), and the doorstop
    /// config all sit alongside the game.
    /// </summary>
    public static bool BepInExInstalled(string gameDir)
    {
        string[] markers =
        {
            Path.Combine(gameDir, "BepInEx", "core"),
            Path.Combine(gameDir, "winhttp.dll"),
            Path.Combine(gameDir, "doorstop_config.ini"),
        };
        return markers.All(m => Directory.Exists(m) || File.Exists(m));
    }

    /// <summary>Whether any Steam process is currently running.</summary>
    public static bool SteamIsRunning()
    {
        try
        {
            return System.Diagnostics.Process.GetProcessesByName("steam").Length > 0;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Whether Valheim itself is currently running — the config editor
    /// polls this to show a "changes apply next launch" notice, since the
    /// game only reads <c>.cfg</c> files at startup. Mirrors the macOS
    /// reference implementation's <c>pgrep -x Valheim</c> check.
    /// </summary>
    public static bool ValheimIsRunning()
    {
        try
        {
            return System.Diagnostics.Process.GetProcessesByName("valheim").Length > 0;
        }
        catch
        {
            return false;
        }
    }

    private static string? TryReadAllText(string path)
    {
        try
        {
            return File.Exists(path) ? File.ReadAllText(path) : null;
        }
        catch
        {
            return null;
        }
    }
}
